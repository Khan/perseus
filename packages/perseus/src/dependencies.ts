import * as React from "react";

import type {PerseusDependencies, PerseusDependenciesV2} from "./types";

let _dependencies: PerseusDependencies | null | undefined = null;

export const setDependencies = (dependencies: PerseusDependencies) => {
    _dependencies = dependencies;
};

export const getDependencies = (): PerseusDependencies => {
    if (_dependencies) {
        return _dependencies;
    }

    throw new Error(
        [
            "Perseus has not been provided required dependencies.",
            "setDependencies(dependencies) must be called first.",
            "Make sure Perseus is being imported from javascript/perseus/perseus.js.",
        ].join("\n"),
    );
};

export type DependencyProps = Partial<
    React.ContextType<typeof DependenciesContext>
>;

export const DependenciesContext = React.createContext<PerseusDependenciesV2>({
    analytics: {onAnalyticsEvent: async () => {}},
});

/**
 * useDependencies provides easy access to the current Perseus dependencies.
 */
export const useDependencies = (): PerseusDependenciesV2 => {
    const deps = React.useContext(DependenciesContext);

    return deps;
};
