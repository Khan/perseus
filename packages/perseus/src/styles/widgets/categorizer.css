.categorizer-container {
    margin-top: 20px;
}
.categorizer-container div.paragraph {
    margin: 10px 0px;
}
.categorizer-container .category {
    text-align: center;
}
.categorizer-container table {
    min-width: 0;
}

/* TODO(benkomalo): ughhh. kill or move this out of here :(
   Styles specifically for the mobile native apps */
body.mobile
    .categorizer-container
    td.category
    input[type="radio"]:checked
    + span:before {
    color: #1c758a;
}
body.mobile
    .categorizer-container
    td.category
    input[type="radio"]
    + span:active:before {
    color: #666;
    content: "\f111"; /* fa-circle */
}
