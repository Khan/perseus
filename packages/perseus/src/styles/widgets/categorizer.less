.categorizer-container {
    margin-top: 20px;

    div.paragraph {
        margin: 10px 0px;
    }

    .category {
        text-align: center;
    }

    table {
        min-width: 0;
    }

    label {
        .above-scratchpad;
    }
}


// TODO(benkomalo): ughhh. kill or move this out of here :(
// Styles specifically for the mobile native apps
body.mobile {
    .categorizer-container {
        td.category {
            input[type="radio"] {
                &:checked + span:before {
                    color: @blue;
                }

                & + span:active:before {
                    color: #666;
                    content: "\f111";  // fa-circle
                }
            }
        }
    }
}
