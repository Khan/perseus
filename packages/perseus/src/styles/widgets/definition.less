@import "../variables.less";
@import "../util.less";

// HACK(michaelpolyak): As the Definition widget may be rendered within an
// Article or an Exercise, we need to support the "legacy" way of styling for
// the specific responsive rendering context. In order to match the surrounding
// Perseus content styling, we copy and apply styles defined in
// `perseus-renderer.less`, `articles.less` (both of which are now moved into
// Perseus).

// Desktop exercise styles
.framework-perseus:not(.perseus-article):not(.perseus-mobile) {
    .perseus-widget-definition {
        .legacy-typography;
    }
}

// Mobile exercise styles
.framework-perseus:not(.perseus-article).perseus-mobile {
    .perseus-widget-definition {
        @desktop-width-min: 1200px;
        @tablet-width-max: (@desktop-width-min - 1);

        // Phone styles
        @media (max-width: @pure-sm-max) {
            font-size: 18px;
            line-height: 1.4;
        }

        // Tablet styles
        @media (min-width: @pure-md-min) and (max-width: @tablet-width-max) {
            font-size: 20px;
            line-height: 1.5;
        }

        // "Desktop" (resolution) styles (while rendering on a mobile device,
        // ie. `this.props.apiOptions.isMobile` is `true`)
        @media (min-width: @desktop-width-min) {
            font-size: 22px;
            line-height: 1.4;
        }
    }
}

// Default article styles
.bibliotron-article.framework-perseus.perseus-article {
    .perseus-widget-definition {
        font-size: 18px;
        line-height: 1.4;
    }
}

// Desktop article styles
.bibliotron-article.framework-perseus.perseus-article:not(.perseus-mobile) {
    .perseus-widget-definition {
        font-size: 20px;
        line-height: 30px;
    }
}

// Mobile article styles
.bibliotron-article.framework-perseus.perseus-article.perseus-mobile {
    .perseus-widget-definition {
        @desktop-width-min: 1200px;
        @tablet-width-max: (@desktop-width-min - 1);

        // Phone styles
        @media (max-width: @pure-sm-max) {
            font-size: 18px;
            line-height: 1.4;
        }

        // Tablet styles
        @media (min-width: @pure-md-min) and (max-width: @tablet-width-max) {
            font-size: 20px;
            line-height: 1.5;
        }

        // "Desktop" (resolution) styles (while rendering on a mobile device,
        // ie. `this.props.apiOptions.isMobile` is `true`)
        @media (min-width: @desktop-width-min) {
            font-size: 22px;
            line-height: 1.4;
        }
    }
}
