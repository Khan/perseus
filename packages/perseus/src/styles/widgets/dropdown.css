.perseus-widget-dropdown {
    position: relative;
}
