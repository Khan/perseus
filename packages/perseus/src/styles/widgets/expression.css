.perseus-widget-expression {
    position: relative;
}
.perseus-widget-expression > span {
    display: inline-block;
    vertical-align: middle;
}
.perseus-widget-expression .perseus-formats-tooltip {
    width: 190px;
}
