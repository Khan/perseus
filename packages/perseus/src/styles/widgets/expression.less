.perseus-widget-expression {
    & > span, .error-tooltip {
        display: inline-block;
        vertical-align: middle;
    }

    position: relative;

    .error-tooltip {
        position: absolute;
        right: 6px;
        top: -2px;
    }

    .error-icon {
        color: #fcc335;
        cursor: pointer;
        font-size: 20px;
        .perseus-interactive;
    }

    .error-text {
        background-color: #fff;
        padding: 5px;
        width: 210px;
    }

    &.show-error-tooltip .perseus-math-input.mq-editable-field.mq-math-mode > .mq-root-block {
        padding-right: 25px;
    }

    .perseus-formats-tooltip {
        width: 190px;
    }
}

#answer_area .perseus-widget-expression {
    // Super hacky overrides to make the error tooltip fit into the answer area
    // by putting it on top of the MathQuill input.

    // TODO(alex): Implement arrows coming from the middle of the tooltip side,
    // then use some of those here; e.g. align to left middle

    .perseus-math-input.mq-editable-field.mq-math-mode {
        // Make the input wider so that error message never scrolls left
        min-width: 130px;
    }

    .error-tooltip {
        .error-text-container {
            // Move the tooltip to the left of the error icon
            left: -125px !important;
            top: -17px !important;
        }

        .error-text {
            // Make the error message much smaller
            font-size: 12px;
            width: 90px;
        }

        .tooltipContainer > div:first-child {
            // Hide the little triangle "arrow" created by the tooltip
            // (If it weren't in a corner, we wouldn't have to hide it)
            visibility: hidden !important;
        }
    }
}

// Old expression widget
.perseus-widget-expression-old {
    &,
    > .output,
    > .output > .tex,
    > .output > .placeholder {
        display: block;
    }
}

.perseus-widget-expression-old input,
#answer_area .perseus-widget-expression-old input {
    direction: ltr;
    border: 1px solid #a4a4a4;
    border-radius: 5px;
    box-sizing: border-box;
    font-size: 14px;
    margin-bottom: 5px;
    max-width: 240px;
    padding: 6px;
    width: 100%;
}

.perseus-widget-expression-old > .output {
    background: #f2f2f2;
    border-radius: 5px;
    padding: 10px;
    margin: 10px 0;
}

#answer_area .perseus-widget-expression-old > .output {
    background: #f7f7f7;
}

.perseus-widget-expression-old > .output > .tex {
    overflow-x: auto;
    padding: 5px;
}

.perseus-widget-expression-old > .output > .placeholder {
    position: relative;
    height: 40px;
    overflow-y: hidden;
}

.perseus-widget-expression-old > .output > .placeholder  > .error {
    -webkit-border-radius: 10px;
    -moz-border-radius: 10px;
    border-radius: 10px;
    background: #f7f7f7;
    border: 1px solid #ddd;
    box-shadow: 2px 2px 5px rgba(0,0,0,0.5);
    color: #000;
    display: none;
    font-weight: normal;
    min-height: 22px;
    width: 168px;
    position: absolute;
    top: 0px;
    left: 40px;
    margin: auto;
}

.perseus-widget-expression-old > .output > .placeholder > .error > .buddy {
    background-image: url(/images/perseus/error-buddy.png);
    background-size: 100%;
    height: 36px;
    width: 40px;
    position: absolute;
    top: -3px;
    left: -42px;
}

.perseus-widget-expression-old > .output > .placeholder > .error > .message {
    font: 12px @baseFontFamily;
    line-height: 1.4em;
    margin: 0px 4px;
}
