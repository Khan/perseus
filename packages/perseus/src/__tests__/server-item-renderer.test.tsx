import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {within, render, screen} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as React from "react";
import "@testing-library/jest-dom"; // Imports custom matchers

import {
    testDependencies,
    testDependenciesV2,
} from "../../../../testing/test-dependencies";
import {
    itemWithInput,
    itemWithLintingError,
    mockedItem,
} from "../__testdata__/server-item-renderer.testdata";
import * as Dependencies from "../dependencies";
import WrappedServerItemRenderer, {
    ServerItemRenderer,
} from "../server-item-renderer";
import {registerWidget} from "../widgets";
import InputNumberExport from "../widgets/input-number";
import RadioWidgetExport from "../widgets/radio";

import MockAssetLoadingWidgetExport, {
    mockedAssetItem,
} from "./mock-asset-loading-widget";
import MockWidgetExport from "./mock-widget";

import type {MockAssetLoadingWidget} from "./mock-asset-loading-widget";
import type {PerseusItem} from "../perseus-types";
import type {APIOptions} from "../types";
import type {KeypadAPI} from "@khanacademy/math-input";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

// This looks alot like `widgets/__tests__/renderQuestion.jsx', except we use
// the ServerItemRenderer instead of Renderer
const renderQuestion = (
    question: PerseusItem,
    apiOptions: APIOptions = Object.freeze({}),
    optionalProps: Partial<
        PropsFor<typeof WrappedServerItemRenderer>
    > = Object.freeze({}),
): {
    container: HTMLElement;
    renderer: ServerItemRenderer;
} => {
    let renderer: ServerItemRenderer | null = null;

    const {container} = render(
        <RenderStateRoot>
            <WrappedServerItemRenderer
                ref={(node) => (renderer = node)}
                apiOptions={apiOptions}
                item={question}
                problemNum={0}
                reviewMode={false}
                dependencies={testDependenciesV2}
                {...optionalProps}
            />
        </RenderStateRoot>,
    );
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    return {container, renderer};
};

describe("server item renderer", () => {
    beforeAll(() => {
        registerWidget("input-number", InputNumberExport);
        registerWidget("radio", RadioWidgetExport);
        registerWidget("mock-widget", MockWidgetExport);
    });

    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    afterEach(() => {
        // The Renderer uses a timer to wait for widgets to complete rendering.
        // If we don't spin the timers here, then the timer fires in the test
        // _after_ and breaks it because we do setState() in the callback,
        // and by that point the component has been unmounted.
        jest.runOnlyPendingTimers();
    });

    it("should snapshot", () => {
        // Arrange and Act
        const {container} = renderQuestion({
            ...itemWithInput,
            hints: [
                {content: "Hint #1", images: {}, widgets: {}},
                {content: "Hint #2", images: {}, widgets: {}},
                {content: "Hint #3", images: {}, widgets: {}},
            ],
        });

        // Assert
        expect(container).toMatchSnapshot("initial render");
    });

    it("should render the content", () => {
        // Arrange and Act
        renderQuestion(itemWithInput);

        // Assert
        expect(screen.getByRole("textbox")).toBeVisible();
    });

    it("should be invalid if no input provided", () => {
        // Arrange
        const {renderer} = renderQuestion(itemWithInput);

        // Act
        const score = renderer.scoreInput();

        // Assert
        expect(score.correct).toBe(false);
        expect(score.empty).toBe(true);
    });

    it("should be answerable", () => {
        // Arrange
        const {renderer} = renderQuestion(itemWithInput);
        userEvent.type(screen.getByRole("textbox"), "-42");

        // Act
        const score = renderer.scoreInput();

        // Assert
        expect(score.correct).toBe(true);
        expect(score.empty).toBe(false);
    });

    it("calls onInteraction callback", () => {
        // Arrange
        const interactionCallback = jest.fn();
        renderQuestion(itemWithInput, {
            interactionCallback,
        });

        // Act
        userEvent.type(screen.getByRole("textbox"), "-42");
        jest.runOnlyPendingTimers(); // Renderer uses setTimeout setting widget props

        // Assert
        expect(interactionCallback).toHaveBeenCalled();
    });

    it("should set the input value for a widget", () => {
        // Arrange
        const focus = jest.fn();
        const {renderer} = renderQuestion(itemWithInput);

        // Act
        renderer.setInputValue(["input-number 1"], "99", focus);

        // Assert
        expect(
            renderer.scoreWidgets()["input-number 1"].guess.currentValue,
        ).toBe("99");
    });

    it("should return the DOM node for the requested focus path", async () => {
        // Arrange
        const {renderer} = renderQuestion(itemWithInput);

        // Act
        const node = renderer.getDOMNodeForPath(["input-number 1"]);

        // Assert
        // @ts-expect-error - TS2345 - Argument of type 'Element | Text | null | undefined' is not assignable to parameter of type 'HTMLElement'.
        expect(await within(node).findAllByRole("textbox")).toHaveLength(1);
    });

    it("should return the grammar type for the requested focus path", () => {
        // Arrange
        const {renderer} = renderQuestion(itemWithInput);

        // Act
        const grammarType = renderer.getGrammarTypeForPath(["input-number 1"]);

        // Assert
        expect(grammarType).toBe("number");
    });

    it("should return the number of hints available", () => {
        // Arrange
        const {renderer} = renderQuestion({
            ...itemWithInput,
            hints: [
                {content: "Hint #1", images: {}, widgets: {}},
                {content: "Hint #2", images: {}, widgets: {}},
                {content: "Hint #3", images: {}, widgets: {}},
            ],
        });

        // Act
        const numHints = renderer.getNumHints();

        // Assert
        expect(numHints).toBe(3);
    });

    it("should return all widget ids", () => {
        // Arrange
        const {renderer} = renderQuestion(mockedItem);

        // Act
        const widgetIds = renderer.getWidgetIds();

        // Assert
        expect(widgetIds).toStrictEqual([
            "mock-widget 1",
            "mock-widget 2",
            "mock-widget 3",
            "mock-widget 4",
        ]);
    });

    it("should call the answerable callback when no widgets are empty", () => {
        // Arrange
        const answerableCallback = jest.fn();
        const {rerender} = render(
            <RenderStateRoot>
                <ServerItemRenderer
                    apiOptions={{
                        answerableCallback,
                    }}
                    item={itemWithInput}
                    problemNum={0}
                    reviewMode={false}
                    dependencies={testDependenciesV2}
                />
            </RenderStateRoot>,
        );
        userEvent.type(screen.getByRole("textbox"), "-42");

        // Act
        rerender(
            <RenderStateRoot>
                <ServerItemRenderer
                    apiOptions={{
                        answerableCallback,
                    }}
                    item={itemWithInput}
                    problemNum={1} // to force componentDidUpdate
                    reviewMode={false}
                    dependencies={testDependenciesV2}
                />
                ,
            </RenderStateRoot>,
        );

        // Assert
        expect(answerableCallback).toHaveBeenCalledWith(true);
    });

    it("should call the onRendered callback when all assets loaded", () => {
        // This is an involved test. We create a mock widget that gives us
        // access to the setAssetStatus function that is passed down by the
        // render tree created.
        // Finally we re-render and poke the asset status to loaded. At that
        // everything is loaded.

        // Arrange
        registerWidget("mocked-asset-widget", MockAssetLoadingWidgetExport);

        const onRendered = jest.fn();
        let renderer: ServerItemRenderer | null | undefined;
        const {rerender} = render(
            <RenderStateRoot>
                <ServerItemRenderer
                    ref={(component) => (renderer = component)}
                    item={mockedAssetItem}
                    problemNum={0}
                    reviewMode={false}
                    onRendered={onRendered}
                    dependencies={testDependenciesV2}
                />
            </RenderStateRoot>,
        );
        if (renderer == null) {
            throw new Error("Renderer failed to render.");
        }

        const mockedWidget = renderer.questionRenderer.getWidgetInstance(
            "mocked-asset-widget 1",
        );
        if (mockedWidget == null) {
            throw new Error("Couldn't find mocked widget!");
        }

        rerender(
            <RenderStateRoot>
                <ServerItemRenderer
                    item={mockedAssetItem}
                    problemNum={1}
                    reviewMode={false}
                    onRendered={onRendered}
                    dependencies={testDependenciesV2}
                />
            </RenderStateRoot>,
        );

        // Act
        // setAssetStatus() is not part of the Widget interface, it's specific
        // this test.
        // @ts-expect-error - TS2352 - Conversion of type 'Widget' to type 'MockAssetLoadingWidget' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
        const widget = mockedWidget as MockAssetLoadingWidget;
        widget.setAssetStatus("ABC", true);

        // Assert
        expect(onRendered).toHaveBeenCalledWith(true);
    });

    describe("focus management", () => {
        it("calls onFocusChange when focusing the renderer", () => {
            // Arranged
            const onFocusChange = jest.fn();
            const {renderer} = renderQuestion(itemWithInput, {
                onFocusChange,
            });

            // Act
            const gotFocus = renderer.focus();

            // We have some async processes that need to be resolved here
            jest.runAllTimers();

            // Assert
            expect(gotFocus).toBeTrue();
            expect(onFocusChange).toHaveBeenCalledWith(
                ["input-number 1"],
                null,
                0,
                expect.any(Object),
            );
        });

        it("activates the keypadElement when focusing the renderer on mobile", () => {
            // Arranged
            const onFocusChange = jest.fn();
            const keypadElementDOMNode = document.createElement("div");

            // We need to mock the getBoundingClientRect() method for our
            // onFocusChange() callback to work properly.
            keypadElementDOMNode.getBoundingClientRect = () =>
                ({
                    height: 250,
                } as DOMRect);

            const keypadElement: KeypadAPI = {
                getDOMNode: jest
                    .fn()
                    .mockImplementation(() => keypadElementDOMNode),
                activate: jest.fn(),
                dismiss: jest.fn(),
                configure: jest.fn(),
                setCursor: jest.fn(),
                setKeyHandler: jest.fn(),
            };
            const {renderer} = renderQuestion(
                itemWithInput,
                {onFocusChange, isMobile: true},
                {keypadElement},
            );

            // Act
            const gotFocus = renderer.focus();

            // We have some async processes that need to be resolved here
            jest.runAllTimers();

            // Assert
            expect(keypadElement.activate).toHaveBeenCalled();
            expect(gotFocus).toBeTrue();
            expect(onFocusChange).toHaveBeenCalledWith(
                ["input-number 1"],
                null,
                250,
                expect.any(Object),
            );
        });

        it("calls onFocusChange when blurring the renderer", () => {
            // Arrange
            const onFocusChange = jest.fn();
            const {renderer} = renderQuestion(itemWithInput, {
                onFocusChange,
            });
            renderer.focus();

            // Act
            renderer.blur();

            // We have some async processes that need to be resolved here
            jest.runAllTimers();

            // Assert
            expect(onFocusChange).toHaveBeenCalledTimes(2);
            expect(onFocusChange).toHaveBeenLastCalledWith(
                null,
                ["input-number 1"],
                0,
                null,
            );
        });

        it("dismisses the keypadElement when blurring the renderer on mobile", () => {
            // Arranged
            const onFocusChange = jest.fn();
            const keypadElementDOMNode = document.createElement("div");

            // We need to mock the getBoundingClientRect() method for our
            // onFocusChange() callback to work properly.
            keypadElementDOMNode.getBoundingClientRect = () =>
                ({
                    height: 250,
                } as DOMRect);

            const keypadElement: KeypadAPI = {
                getDOMNode: jest
                    .fn()
                    .mockImplementation(() => keypadElementDOMNode),
                activate: jest.fn(),
                dismiss: jest.fn(),
                configure: jest.fn(),
                setCursor: jest.fn(),
                setKeyHandler: jest.fn(),
            };
            const {renderer} = renderQuestion(
                itemWithInput,
                {onFocusChange, isMobile: true},
                {keypadElement},
            );
            renderer.focus();

            // Act
            renderer.blur();

            // We have some async processes that need to be resolved here
            jest.runAllTimers();

            // Assert
            expect(keypadElement.dismiss).toHaveBeenCalled();
            expect(onFocusChange).toHaveBeenCalledTimes(2);
            expect(onFocusChange).toHaveBeenLastCalledWith(
                null,
                ["input-number 1"],
                0,
                null,
            );
        });

        it("should focus the widget requested in focusPath()", () => {
            // Arrange
            const onFocusChange = jest.fn();
            const {renderer} = renderQuestion(itemWithInput, {
                onFocusChange,
            });

            // Act
            renderer.focusPath(["input-number 1"]);

            // We have some async processes that need to be resolved here
            jest.runAllTimers();

            // Assert
            expect(onFocusChange).toHaveBeenCalledWith(
                ["input-number 1"],
                null,
                0,
                expect.any(Object),
            );
        });
    });

    describe("state serialization", () => {
        it("should serialize the current state", () => {
            // Arrange
            const {renderer} = renderQuestion({
                ...itemWithInput,
                hints: [
                    {content: "Hint #1", images: {}, widgets: {}},
                    {content: "Hint #2", images: {}, widgets: {}},
                    {content: "Hint #3", images: {}, widgets: {}},
                ],
            });
            userEvent.type(screen.getByRole("textbox"), "-42");

            // Act
            const state = renderer.getSerializedState();

            // Assert
            expect(state).toMatchInlineSnapshot(`
                {
                  "hints": [
                    {},
                    {},
                    {},
                  ],
                  "question": {
                    "input-number 1": {
                      "answerType": "number",
                      "currentValue": "-42",
                      "rightAlign": undefined,
                      "simplify": "required",
                      "size": "normal",
                    },
                  },
                }
            `);
        });

        it("should restore serialized state", () => {
            // Arrange
            const callback = jest.fn();
            const {renderer} = renderQuestion(itemWithInput);

            // Act
            renderer.restoreSerializedState(
                {
                    hints: [{}, {}, {}],
                    question: {
                        "input-number 1": {
                            answerType: "number",
                            currentValue: "-42",
                            rightAlign: undefined,
                            simplify: "required",
                            size: "normal",
                        },
                    },
                },
                callback,
            );
            jest.runOnlyPendingTimers();

            // Assert
            expect(callback).toHaveBeenCalled();
            expect(screen.getByRole("textbox")).toHaveValue("-42");
        });
    });

    describe("content editing", () => {
        it("shouldn't show linting errors when highlightLint is false", () => {
            // Arrange and Act
            renderQuestion(itemWithLintingError, undefined, {
                linterContext: {
                    contentType: "exercise",
                    highlightLint: false,
                    paths: [],
                    stack: [],
                },
            });

            expect(
                screen.queryByText("Don't use level-1 headings", {
                    exact: false,
                }),
            ).not.toBeInTheDocument();
        });

        it("should show linting errors when highlightLint is true", () => {
            // Arrange and Act
            renderQuestion(itemWithLintingError, undefined, {
                linterContext: {
                    contentType: "exercise",
                    highlightLint: true,
                    paths: [],
                    stack: [],
                },
            });

            expect(
                screen.getByText("Don't use level-1 headings", {exact: false}),
            ).toBeInTheDocument();
        });
    });
});
