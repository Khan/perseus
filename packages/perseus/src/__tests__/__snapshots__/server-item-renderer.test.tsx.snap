// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server item renderer should snapshot: initial render 1`] = `
<div>
  <div>
    <div>
      <div
        class="perseus-renderer perseus-renderer-responsive"
      >
        <div
          class="paragraph"
          data-perseus-paragraph-index="0"
        >
          <div
            class="paragraph"
          >
            Enter the number 
            <span
              style="white-space: nowrap;"
            >
              <span />
              <span
                class="mock-TeX"
              />
              <span />
            </span>
            -42
            <span
              style="white-space: nowrap;"
            >
              <span />
              <span
                class="mock-TeX"
              />
              <span />
            </span>
             in the box: 
            <div
              class="perseus-widget-container widget-nohighlight widget-inline-block"
            >
              <span>
                <div>
                  <input
                    autocapitalize="off"
                    autocomplete="off"
                    autocorrect="off"
                    class="input_1t4b01j-o_O-LabelMedium_1rew30o-o_O-default_1r2dwp3-o_O-default_1gznapd-o_O-leftAlign_7jg4r4"
                    id="input-with-examples-aW5wdXQtbnVtYmVyIDE"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  style="position: relative; height: 0px; display: none;"
                >
                  <div
                    class="tooltipContainer"
                    style="position: absolute; left: 0px;"
                  >
                    <div
                      style="display: block; position: relative; visibility: visible; left: 0px; top: -1px; width: 12px; height: 11px; margin-top: -1px; margin-bottom: -2px; z-index: 10;"
                    >
                      <div
                        style="display: block; height: 0px; width: 0px; position: absolute; left: 0px; top: -1px; border-right: 12px solid transparent; border-bottom: 12px solid #ccc;"
                      />
                      <div
                        style="display: block; height: 0px; width: 0px; position: absolute; left: 1px; top: 1px; border-right: 10px solid transparent; border-bottom: 10px solid white;"
                      />
                    </div>
                    <div
                      class="perseus-formats-tooltip preview-measure"
                      style="position: relative; top: 0px; left: 0px; border: 1px solid #ccc; box-shadow: 0 1px 3px #ccc; z-index: 9;"
                    >
                      <div
                        id="input-with-examples-aW5wdXQtbnVtYmVyIDE"
                      >
                        <div
                          class="perseus-renderer perseus-renderer-responsive"
                        >
                          <div
                            class="paragraph"
                            data-perseus-paragraph-index="0"
                          >
                            <ul>
                              <li>
                                <strong>
                                  Your answer should be
                                </strong>
                              </li>
                              <li>
                                an integer, like 
                                <span
                                  style="white-space: nowrap;"
                                >
                                  <span />
                                  <span
                                    class="mock-TeX"
                                  >
                                    6
                                  </span>
                                  <span />
                                </span>
                              </li>
                              <li>
                                an 
                                <em>
                                  exact
                                </em>
                                 decimal, like 
                                <span
                                  style="white-space: nowrap;"
                                >
                                  <span />
                                  <span
                                    class="mock-TeX"
                                  >
                                    0.75
                                  </span>
                                  <span />
                                </span>
                              </li>
                              <li>
                                a 
                                <em>
                                  simplified proper
                                </em>
                                 fraction, like 
                                <span
                                  style="white-space: nowrap;"
                                >
                                  <span />
                                  <span
                                    class="mock-TeX"
                                  >
                                    3/5
                                  </span>
                                  <span />
                                </span>
                              </li>
                              <li>
                                a 
                                <em>
                                  simplified improper
                                </em>
                                 fraction, like 
                                <span
                                  style="white-space: nowrap;"
                                >
                                  <span />
                                  <span
                                    class="mock-TeX"
                                  >
                                    7/4
                                  </span>
                                  <span />
                                </span>
                              </li>
                              <li>
                                a mixed number, like 
                                <span
                                  style="white-space: nowrap;"
                                >
                                  <span />
                                  <span
                                    class="mock-TeX"
                                  >
                                    1\\ 3/4
                                  </span>
                                  <span />
                                </span>
                              </li>
                            </ul>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="hintsContainer_1g61p95"
    >
      <div
        class=""
      >
        <div
          class="hint perseus-hint-renderer"
          tabindex="-1"
        >
          <span
            class="perseus-sr-only"
          >
            Hint #1
          </span>
          <span
            class="perseus-hint-label"
            style="display: block;"
          >
            1 / 3
          </span>
          <div
            class="perseus-renderer perseus-renderer-responsive"
          >
            <div
              class="paragraph"
              data-perseus-paragraph-index="0"
            >
              <div
                class="paragraph"
              >
                Hint #1
              </div>
            </div>
          </div>
        </div>
        <div
          class="hint perseus-hint-renderer"
          tabindex="-1"
        >
          <span
            class="perseus-sr-only"
          >
            Hint #2
          </span>
          <span
            class="perseus-hint-label"
            style="display: block;"
          >
            2 / 3
          </span>
          <div
            class="perseus-renderer perseus-renderer-responsive"
          >
            <div
              class="paragraph"
              data-perseus-paragraph-index="0"
            >
              <div
                class="paragraph"
              >
                Hint #2
              </div>
            </div>
          </div>
        </div>
        <div
          class="hint perseus-hint-renderer last-hint last-rendered"
          tabindex="-1"
        >
          <span
            class="perseus-sr-only"
          >
            Hint #3
          </span>
          <span
            class="perseus-hint-label"
            style="display: block;"
          >
            3 / 3
          </span>
          <div
            class="perseus-renderer perseus-renderer-responsive"
          >
            <div
              class="paragraph"
              data-perseus-paragraph-index="0"
            >
              <div
                class="paragraph"
              >
                Hint #3
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
